functions/src/alchemy_webhook.ts

```diff
@@ -2,16 +2,106 @@ import * as functions from "firebase-functions";
+type Log = {
+  removed: boolean;
+  address: string;
+  data: string;
+  topics: string[];
+};
+
+type RawContract = {
+  rawvalue?: string;
+  address?: string;
+  decimal?: number;
+};
+
+type Activity = {
+  category: string;
+  fromAddress: string;
+  toAddress: string;
+  blockNum: string;
+  value?: number;
+  erc721TokenId?: string;
+  asset?: string;
+  hash: string;
+  typeTraceAddress: string;
+  rawContract?: RawContract;
+  log?: Log;
+};
+
+type AddressEvent = {
+  network: string;
+  activity: Activity[];
+};
+
+type AddressNotification = {
+  webhookId: string;
+  id: string;
+  createdAt: string;
+  type: string;
+  event: AddressEvent;
+};
+
+// async getProfile(channel: string) {
+//   const results = await this.firestore
+//     .collection("profiles")
+//     .where(`${this.provider}.login`, "==", channel)
+//     .get();
+//   if (results.size > 1) {
+//     log.error(
+//       { provider: this.provider, channel },
+//       "duplicate profiles found"
+//     );
+//   }
+//   return results.empty ? null : results.docs[0];
+// }
+
 // TODO: webhook receiver from alchemy
 export const alchemyWebhook = functions.https.onRequest(async (req, res) => {
-  // const body = req.body;
-  // const event = body.event;
-  // const alchemyWebhookId = body.webhookId;
-  // const network = event.network;
+  const body = req.body;
+  const notification: AddressNotification = body as AddressNotification;
+  for (const activity of notification.event.activity) {
+    // look for userId that associated with this address
+    await admin
+      .firestore()
+      .collection("realtimecash")
+      .where("address", "==", activity.toAddress)
+      .get()
+      .then(async (snapshot) => {
+        // get channelId for each userId
+        const profile = await admin
+          .firestore()
+          .collection("profiles")
+          .where("userId", "==", snapshot.docs[0].data().userId)
+          .get();
+        
+        const channelId = profile.docs[0].data().id;
+        const login = profile.docs[0].data().login;
+        const displayName = profile.docs[0].data().displayName;
 
-  // // list of activities
-  // const activitities = event.activity;
-  res.status(200).send();
+        // storing donation respoonses in realtimecash collection
+        await admin.firestore().collection("messages").add({
+          channelId: channelId,
+          broadcaster_user_id: channelId,
+          broadcaster_user_login: login,
+          broadcaster_user_name: displayName,
+          userId: snapshot.docs[0].data().userId,
+          webhookId: notification.webhookId,
+          id: notification.id,
+          createdAt: notification.createdAt,
+          type: "realtimecash.donation",
+          notificationType: notification.type,
+          activity: activity,
+        });
+      })
+      .catch((error) => {
+        // no userId found
+        console.log(error);
+        res.status(400).send("No userId found");
+        return;
+      });
```

Comment:
```suggestion
```

If we return a 400, the webhook will keep retrying. It's fine to return 200 since the profile missing is a non-recoverable error.